cpp/include/cudf/ast/operators.hpp

```diff
@@ -0,0 +1,1159 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cmath>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <simt/type_traits>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace cudf {
+
+namespace ast {
+
+/**
+ * @brief Enum of supported operators.
+ *
+ */
+enum class ast_operator {
+  // Binary operators
+  ADD,            ///< operator +
+  SUB,            ///< operator -
+  MUL,            ///< operator *
+  DIV,            ///< operator / using common type of lhs and rhs
+  TRUE_DIV,       ///< operator / after promoting type to floating point
+  FLOOR_DIV,      ///< operator / after promoting to 64 bit floating point and then
+                  ///< flooring the result
+  MOD,            ///< operator %
+  PYMOD,          ///< operator % but following python's sign rules for negatives
+  POW,            ///< lhs ^ rhs
+  EQUAL,          ///< operator ==
+  NOT_EQUAL,      ///< operator !=
+  LESS,           ///< operator <
+  GREATER,        ///< operator >
+  LESS_EQUAL,     ///< operator <=
+  GREATER_EQUAL,  ///< operator >=
+  BITWISE_AND,    ///< operator &
+  BITWISE_OR,     ///< operator |
+  BITWISE_XOR,    ///< operator ^
+  LOGICAL_AND,    ///< operator &&
+  LOGICAL_OR,     ///< operator ||
+  // Unary operators
+  IDENTITY,    ///< Identity function
+  SIN,         ///< Trigonometric sine
+  COS,         ///< Trigonometric cosine
+  TAN,         ///< Trigonometric tangent
+  ARCSIN,      ///< Trigonometric sine inverse
+  ARCCOS,      ///< Trigonometric cosine inverse
+  ARCTAN,      ///< Trigonometric tangent inverse
+  SINH,        ///< Hyperbolic sine
+  COSH,        ///< Hyperbolic cosine
+  TANH,        ///< Hyperbolic tangent
+  ARCSINH,     ///< Hyperbolic sine inverse
+  ARCCOSH,     ///< Hyperbolic cosine inverse
+  ARCTANH,     ///< Hyperbolic tangent inverse
+  EXP,         ///< Exponential (base e, Euler number)
+  LOG,         ///< Natural Logarithm (base e)
+  SQRT,        ///< Square-root (x^0.5)
+  CBRT,        ///< Cube-root (x^(1.0/3))
+  CEIL,        ///< Smallest integer value not less than arg
+  FLOOR,       ///< largest integer value not greater than arg
+  ABS,         ///< Absolute value
+  RINT,        ///< Rounds the floating-point argument arg to an integer value
+  BIT_INVERT,  ///< Bitwise Not (~)
+  NOT,         ///< Logical Not (!)
+  // Other operators -- not implemented
```

Comment:
Do we need to keep the commented out stuff?

Reply:
I removed the commented operators.